src/guest_memory.rs

```diff
@@ -0,0 +1,237 @@
+            Error::ReadFromMemory(_) => write!(f, "couldn't read from memory"),
+            Error::WriteToTarget(_) => write!(f, "couldn't write to target"),
+            Error::ReadFromSource(_) => write!(f, "couldn't read from source"),
+            Error::ShortWrite {
+                expected,
+                completed,
+            } => write!(
+                f,
+                "incomplete write of {} instead of {} bytes",
+                completed, expected,
+            ),
+            Error::ShortRead {
+                expected,
+                completed,
+            } => write!(
+                f,
+                "incomplete read of {} instead of {} bytes",
+                completed, expected,
+            ),
+            Error::SystemCallFailed(e) => write!(f, "syscall failed due to {}", e),
+            Error::InvalidBackendAddress => write!(f, "invalid backend address"),
+            Error::InvalidBackendOffset => write!(f, "invalid backend offset"),
+            Error::InvalidBackendOperation => write!(f, "invalid backend operation"),
+        }
+    }
+}
+
+/// Represents a guest physical address (GPA).
+///
+/// Notes:
+/// - On ARM64, a 32-bit hypervisor may be used to support a 64-bit guest. For simplicity,
+/// u64 is used to store the the raw value no matter the guest a 32-bit or 64-bit virtual machine.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct GuestAddress(pub u64);
+impl_address_ops!(GuestAddress, u64);
+
+/// Type of the raw value stored in a GuestAddress object.
+pub type GuestAddressValue = <GuestAddress as AddressValue>::V;
+
+/// Type to encode offset in the guest physical address space.
+pub type GuestAddressOffset = <GuestAddress as AddressValue>::V;
+
+/// Represents a continuous region of guest physical memory.
+pub trait GuestMemoryRegion: AddressRegion<A = GuestAddress, E = Error> {}
+
+/// Represents a collection of GuestMemoryRegion objects.
+///
+/// The main responsibilities of the GuestMemory trait are:
+/// - hide the detail of accessing guest's physical address.
+/// - map a request address to a GuestMemoryRegion object and relay the request to it.
+/// - handle cases where an access request spanning two or more GuestMemoryRegion objects.
+pub trait GuestMemory: AddressSpace<GuestAddress, Error> {
+    /// Invoke callback `f` to handle data in the address range [addr, addr + count).
+    ///
+    /// The address range [addr, addr + count) may span more than one AddressRegion objects, or
+    /// even has holes within it. So try_access() invokes the callback 'f' for each AddressRegion
+    /// object involved and returns:
+    /// - error code returned by the callback 'f'
+    /// - size of data already handled when encountering the first hole
+    /// - size of data already handled when the whole range has been handled
+    fn try_access<F>(&self, count: usize, addr: GuestAddress, mut f: F) -> Result<usize, Error>
```

Comment:
I'm a bit apprehensive about this function and its uses throughout the PR. I'm mainly concerned about the way `Bytes` is implemented for `T: GuestMemory` (somewhere in a later commit), and the fact that we silently allow reads/writes across multiple regions (even if they have to be adjacent) by default. I think that cases where a read/write spills into another region are erroneous situations more often than not, and the implementation of `Bytes` for `T: GuestMemory` should not touch more than one region. 

To go around this limitation for cases where it actually makes sense, we can offer a helper method like `try_access` that iterates through all relevant regions and invokes some logic passed by the user. What do you think?

Reply:
The main target is to support memory hotplug(hot-adding). A request crossing region boundary may be valid if memory hot-adding is enabled. Providing a helper such as try_access() doesn't solve the issue because the consumer has no way to tell whether a memory region is hot-added or not so it will always use try_access().

Reply:
@jiangliu can we add another feature to vm-memory called "hot-plug" or something of that sort? @alexandruag would that be an acceptable compromise?

Reply:
I don't think it is a matter of compromising; one behavior is right and the other is wrong.

Hotplug is an example, but in general this is just how DMA works, and at its heart a VMM is just emulating hardware. Firecracker does not implement this behavior but it's incorrect. The purpose of rust-vmm should also be to avoid this kind of shortcuts and do things the right way.

Reply:
Thanks for the added context! There are obviously a lot of things that I'm missing, but it's not super clear that one behavior is correct/desirable every single time, whereas the other shouldn't ever make sense for anyone :( 

How about decoupling any particular behavior from the interface and leaving it up the implementation? In this case, we remove the `try_access` entirely from `GuestMemory`, and leave every method that relies on it unimplemented. We can then use `try_access` in the reference implementation, where we might also (at some point) add a parameter/feature/compilation flag that configures whether cross-region access is allowed.

Reply:
How is it useful _not_ to cross regions? IMO the point of this crate is exactly to provide a good implementation for things that are not so obvious, and this qualifies perfectly...